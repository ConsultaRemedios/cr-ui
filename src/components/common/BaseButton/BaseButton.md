### Sizes

```js
  <div>
    <BaseButton size="small">Button small</BaseButton>
    <BaseButton size="medium">Button medium</BaseButton>
    <BaseButton size="large">Button large</BaseButton>
  </div>
```

### Types

```js 
  <div>
    <BaseButton type="success">Button success</BaseButton>
    <BaseButton type="info">Button info</BaseButton>
    <BaseButton type="warning">Button warning</BaseButton>
    <BaseButton type="neutral">Button neutral</BaseButton>
    <BaseButton type="danger">Button danger</BaseButton>
    <BaseButton type="naked">Button naked</BaseButton>
    <BaseButton type="naked-red">Button naked-red</BaseButton>
  </div>
```

### Fill

Fill property forces the button width to fill 100% width of its parent.

```js
<div>
  <div :style="{ marginBottom: '15px' }">
    <BaseButton :fill="true">Example 1</BaseButton>
  </div>
  <div>
    <BaseButton type="neutral" :fill="true">Example 2</BaseButton>
  </div>
</div>
```

### Combining type and size props

```js
  <div>
    <BaseButton type="neutral" size="large">Button neutral large</BaseButton>
    <BaseButton type="success" size="medium">Button success medium</BaseButton>
  </div>
```

### Button as an achor tag

Just add a `path` prop with the url you want to link to as the value.
You can also open the link in a new tab using the `openNewTab` prop as true

```js
  <div>
    <BaseButton type="neutral" path="http://www.google.com">
      Anchor default
    </BaseButton>

    <BaseButton type="success" size="large" path="http://www.google.com" :openNewTab="true">
      Anchor new tab
    </BaseButton>
  </div>
```

### Disabled State

```js
<div>
  <BaseButton :disabled="true">Button neutral</BaseButton>
  <BaseButton :disabled="true" type="naked">Button naked</BaseButton>
  <BaseButton :disabled="true" type="naked-red">Button naked red</BaseButton>
</div>
```

### Inverted option

Invert the background-color and border-color by adding this option

```js
<div>
  <BaseButton :inverted="true" type="success">Success inverted</BaseButton>
  <BaseButton :inverted="true" type="info">Info inverted</BaseButton>
  <BaseButton :inverted="true" type="warning">Warning inverted</BaseButton>
  <BaseButton :inverted="true" type="danger">Danger inverted</BaseButton>
  <BaseButton :inverted="true" type="neutral">Neutral inverted</BaseButton>
</div>
```

### Loading button

Add a behavior of loading to button

```js
const loading = false;

<div>
  <label>
    is Loading:
    <input type="checkbox" @input="loading = !loading"/>
  </label>

  <br/>
  <br/>

  <BaseButton type="success" :isLoading="loading">Loading button</BaseButton>
  <BaseButton type="neutral" :isLoading="loading">Loading button</BaseButton>
</div>
```


### Button with icon

Add a icon in button

```js
  const getIcon = (icons, name) => icons.find((icon) => icon.id === `${name}.icon`);

  <div>
      <BaseButton size="small" :icon="getIcon(icons, 'info')" type="success">Button success</BaseButton>
      <BaseButton size="medium" :icon="getIcon(icons, 'cancel')" type="neutral">Button success</BaseButton>
      <BaseButton size="large" :icon="getIcon(icons, 'change')" type="warning">Button success</BaseButton>
  </div>
```