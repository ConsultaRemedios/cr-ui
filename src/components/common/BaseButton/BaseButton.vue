<template>
  <component :is="element"
    :type="buttonType"
    :disabled="disabled || isLoading"
    :class="[
      $style.button,
      $style[size],
      $style[type],
      {
        [$style.inverted]: inverted,
        [$style.disabled]: disabled,
        [$style.fill]: fill,
        [$style.buttonLoading]: isLoading,
        [$style.buttonShadow]: type !== 'naked' && type !== 'naked-red',
      }
    ]"
    :target="target"
    @click="onClick"
    :href="path"
    v-bind="$attrs"
  >
    <BaseIcon
      v-if="isLoading"
      :id="loaderIcon.id"
      :class="$style.loaderIcon"
    />
    <BaseIcon
      v-if="icon"
      :id="icon.id"
      :class="$style.iconButton"
    />
    <span :class="[{[$style.textLoading]: isLoading}]">
      <slot></slot>
    </span>
  </component>
</template>

<script>
  import BaseIcon from '../BaseIcon';
  import loaderIcon from '../../../icons/loader.icon.svg';

  export default {
    inheritAttrs: false,
    name: 'BaseButton',
    props: {
      buttonType: {
        type: String,
        validator(value) {
          return ['button', 'submit'].includes(value);
        },
      },

      size: {
        type: String,
        default: 'medium',
        validator(value) {
          return ['small', 'medium', 'large'].includes(value);
        },
      },

      type: {
        type: String,
        default: 'success',
        validator(value) {
          return ['success', 'info', 'warning', 'danger', 'neutral', 'naked', 'naked-red'].includes(value);
        },
      },

      disabled: {
        type: Boolean,
        default: false,
      },

      fill: {
        type: Boolean,
        default: false,
      },

      path: {
        type: String,
      },

      openNewTab: {
        type: Boolean,
        default: false,
      },

      inverted: {
        type: Boolean,
        default: false,
      },

      isLoading: {
        type: Boolean,
        default: false,
      },
      icon: {
        type: Object,
      },
    },

    computed: {
      element() {
        return this.path ? 'a' : 'button';
      },

      target() {
        return this.openNewTab ? '_blank' : null;
      },

      loaderIcon() {
        return loaderIcon;
      },
    },

    methods: {
      onClick() {
        this.$emit('click');
      },
    },

    components: {
      BaseIcon,
    },
  };
</script>

<style module>
  @value media-md from './../../../styles/variables.css';

  .button > span {
    display: inline-block;
    line-height: 24px;
  }

  .button {
    border-radius: 4px;
    cursor: pointer;
    text-decoration: none;
    display: inline-block;
    text-align: center;
    transition: background-color 0.25s ease-out,
                border-color 0.25s ease-out,
                color 0.25s ease-out;
  }

  .buttonShadow {
    box-shadow: 0 1px 6px 0 rgba(0,0,0,0.15);
  }

  .button:hover {
    text-decoration: none;
  }

  .button:focus {
    outline: none;
  }

  /* Sizes */
  .small {
    font-size: 14px;
    font-weight: 600;
    padding: 1px 8px;
  }

  .medium {
    padding: 7px 18px;
    font-size: 14px;
    font-weight: 600;
  }

  .large {
    font-size: 14px;
    font-weight: 600;
    padding: 10px 20px;
  }

  /* Fill */
  .fill {
    width: 100%;
  }

  /* Types */
  .success {
    background-color: #1BB738;
    border: 1px solid #1BB738;
    color: #FFF;
  }

  .success:focus,
  .success:hover {
    background-color: #16942D;
    border-color: #16942D;
    color: #FFF;
  }

  .info {
    background-color: #3DC2C2;
    border-color: #3DC2C2;
    border-style: solid;
    color: #FFF;
  }

  .info:focus,
  .info:hover {
    background-color: #169998;
    border-color: #169998;
    color: #FFF;
  }

  .warning {
    background-color: #F9AD3D;
    border-color: #F9AD3D;
    border-style: solid;
    color: #FFF;
  }

  .warning:focus,
  .warning:hover {
    background-color: #CA8C31;
    border-color: #CA8C31;
    color: #FFF;
  }

  .neutral {
    background-color: #FFF;
    border: 1px solid #D9D9D9;
    color: #333333;
  }

  .neutral:focus,
  .neutral:hover {
    background-color: #F8F8F8;
    border-color: #D9D9D9;
    color: #333333;
  }

  .danger {
    background-color: #D30B3F;
    border-color: #D30B3F;
    border-style: solid;
    color: #FFF;
  }

  .danger:focus,
  .danger:hover {
    background-color: #AB0933;
    border-color: #AB0933;
    color: #FFF;
  }

  .naked {
    border: 0;
    background: transparent;
    color: #00AAE5;
  }

  .naked:focus,
  .naked:hover {
    color: #007CA7;
  }

  .naked-red {
    composes: naked;
    color: #D30B3F;
  }

  .naked-red:hover {
    color: #AB0933;
  }

  /* States */
  .disabled,
  .disabled:hover,
  .disabled:focus {
    cursor: default;
    background-color: #EEE;
    border-color: #EEE;
    color: #CCC;
  }

  .inverted.success {
    border: 1px solid #1BB738;
    color: #1BB738;
  }

  .inverted.info {
    border: 1px solid #3DC2C2;
    color: #3DC2C2;
  }

  .inverted.warning {
    border: 1px solid #F9AD3D;
    color: #F9AD3D;
  }

  .inverted.neutral {
    background-color: #D9D9D9;
    border: 1px solid #D9D9D9;
  }

  .inverted.neutral:hover {
    background-color: #333;
    border-color: #333;
  }

  .inverted.danger {
    border: 1px solid #D30B3F;
    color: #D30B3F;
  }

  .inverted {
    background-color: #FFF;
  }

  .inverted:hover {
    color: #FFF;
  }

  /* Combined type and size */
  .naked.medium,
  .naked-red.medium {
    font-size: 14px;
    font-weight: 800;
  }

  .naked.disabled,
  .naked.disabled:hover {
    background-color: transparent;
  }

  .loaderIcon {
    animation: rotate-loader 1.5s infinite;

    position:absolute;
    top: 50%;
    left:0;
    right:0;
    margin-left:auto;
    margin-right:auto;
    transform: translateY(-50%);
  }

  .iconButton {
    margin-right: 4px;
  }

  .buttonLoading {
    position: relative;
  }

  .textLoading {
    opacity: 0;
  }

  @media media-md {
    .large {
      padding-left: 25px;
      padding-right: 25px;
    }
  }

  @keyframes rotate-loader {
    0% {
      transform: translateY(-50%) rotate(0deg);
    }

    100% {
      transform: translateY(-50%) rotate(360deg);
    }
  }
</style>
