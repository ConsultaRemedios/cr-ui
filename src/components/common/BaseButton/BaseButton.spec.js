import { mount } from '@vue/test-utils';
import loaderIcon from '../../../icons/loader.icon.svg';
import BaseButton from './BaseButton.vue';

const mountButton = ({
  buttonType = undefined,
  inverted = false,
  fill = false,
  size = 'medium',
  type = 'success',
  text,
  icon = undefined,
  path,
  atributtes = { },
}) => (
  mount(BaseButton, {
    propsData: {
      fill,
      size,
      type,
      inverted,
      buttonType,
      icon,
      path,
    },
    attrs: {
      ...atributtes
    },
    slots: {
      default: text
    },
  })
);

describe('BaseButton component', () => {
  describe('#buttonType prop', () => {
    it('matches the snapshoot when there is not buttonType prop', () => {
      const wrapper = mountButton({ text: 'Button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when buttonType is equal to submit', () => {
      const wrapper = mountButton({ buttonType: 'submit', text: 'Button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when buttonType is equal to button', () => {
      const wrapper = mountButton({ buttonType: 'button', text: 'Button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('is valid if the value is button', () => {
      const wrapper = mountButton({ buttonType: 'button', text: 'Button' });
      expect(wrapper.vm.$options.props.buttonType.validator('button')).toBeTruthy()
    });

    it('is valid if the value is submit', () => {
      const wrapper = mountButton({ buttonType: 'button', text: 'Button' });
      expect(wrapper.vm.$options.props.buttonType.validator('submit')).toBeTruthy()
    });

    it('is not valid if the value is not equal to button or submit', () => {
      const wrapper = mountButton({ text: 'Button' });

      ['asd', 'foo', 'bar', 'zaz', 'woww'].forEach((v) => {
        expect(wrapper.vm.$options.props.buttonType.validator(v)).toBeFalsy();
      });
    });
  });

  describe('#size prop', () => {
    it('matches the snapshoot when the size is "small"', () => {
      const wrapper = mountButton({ size: 'small', text: 'Small button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshot when the size is "medium"', () => {
      const wrapper = mountButton({ size: 'medium', text: 'Medium button' });
      expect(wrapper.element).toMatchSnapshot();
    })

    it('matches the snapshot when the size is "large"', () => {
      const wrapper = mountButton({ size: 'large', text: 'Large button' });
      expect(wrapper.element).toMatchSnapshot();
    })
  });

  describe('#type prop', () => {
    it('matches the snapshoot when the type is "success"', () => {
      const wrapper = mountButton({ type: 'success', text: 'Success button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when the type is "info"', () => {
      const wrapper = mountButton({ type: 'info', text: 'Info button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when the type is "warning"', () => {
      const wrapper = mountButton({ type: 'warning', text: 'Warning button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when the type is "danger"', () => {
      const wrapper = mountButton({ type: 'danger', text: 'Danger button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when the type is "neutral"', () => {
      const wrapper = mountButton({ type: 'neutral', text: 'Neutral button' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshoot when the type is "naked"', () => {
      const wrapper = mountButton({ type: 'naked', text: 'Naked button' });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('#fill prop', () => {
    describe('When fill is true', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ fill: true, text: 'Filled button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When fill is false', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ text: 'Unfilled button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('#inverted prop', () => {
    describe('When inverted is true', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ inverted: true, text: 'Inverted button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When inverted is false', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ text: 'Button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('#isLoading prop', () => {
    describe('When isLoading is true', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ isLoading: true, text: 'Loading button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When isLoading is false', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ text: 'Not Loading button' });
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('#icon prop', () => {
    describe('when set an icon', () => {
      it('it matches the snapshot', () => {
        const wrapper = mountButton({ icon: { icon: loaderIcon }, text: 'Button success' });
        expect(wrapper.element).toMatchSnapshot();
      })
    })
  })

  describe('#rel prop', () => {
    describe('When has rel prop', () => {
      it('matches the snapshot', () => {
        const wrapper = mountButton({ text: 'Link button with rel', path: '/foo', atributtes: {  rel: 'nofollow' } });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('With combined props', () => {
    it('matches the snapshot', () => {
      const wrapper = mountButton({
        size: 'large',
        type: 'warning',
        fill: true,
        text: 'Combined button'
      });

      expect(wrapper.element).toMatchSnapshot();
    })
  });

  describe('When it is disabled', () => {
    it('matches the snapshot', () => {
      const wrapper = mountButton({
        type: 'danger',
        disabled: true,
        text: 'Disabled button'
      });
    });
  });
});
