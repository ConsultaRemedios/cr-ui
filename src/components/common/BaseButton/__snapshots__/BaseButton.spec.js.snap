// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton component #buttonType prop matches the snapshoot when buttonType is equal to button 1`] = `
<button
  class="button medium success buttonShadow"
  type="button"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Button
  </span>
</button>
`;

exports[`BaseButton component #buttonType prop matches the snapshoot when buttonType is equal to submit 1`] = `
<button
  class="button medium success buttonShadow"
  type="submit"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Button
  </span>
</button>
`;

exports[`BaseButton component #buttonType prop matches the snapshoot when there is not buttonType prop 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Button
  </span>
</button>
`;

exports[`BaseButton component #fill prop When fill is false matches the snapshot 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Unfilled button
  </span>
</button>
`;

exports[`BaseButton component #fill prop When fill is true matches the snapshot 1`] = `
<button
  class="button medium success fill buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Filled button
  </span>
</button>
`;

exports[`BaseButton component #icon prop when set an icon it matches the snapshot 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <svg
    class="icon iconButton"
    viewBox="0 0 24 24"
  >
    <use
      xlink:href="#undefined"
    />
  </svg>
   
  <span
    class=""
  >
    Button success
  </span>
</button>
`;

exports[`BaseButton component #inverted prop When inverted is false matches the snapshot 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Button
  </span>
</button>
`;

exports[`BaseButton component #inverted prop When inverted is true matches the snapshot 1`] = `
<button
  class="button medium success inverted buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Inverted button
  </span>
</button>
`;

exports[`BaseButton component #isLoading prop When isLoading is false matches the snapshot 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Not Loading button
  </span>
</button>
`;

exports[`BaseButton component #isLoading prop When isLoading is true matches the snapshot 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Loading button
  </span>
</button>
`;

exports[`BaseButton component #rel prop When has rel prop matches the snapshot 1`] = `
<a
  class="button medium success buttonShadow"
  href="/foo"
  rel="nofollow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Link button with rel
  </span>
</a>
`;

exports[`BaseButton component #size prop matches the snapshoot when the size is "small" 1`] = `
<button
  class="button small success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Small button
  </span>
</button>
`;

exports[`BaseButton component #size prop matches the snapshot when the size is "large" 1`] = `
<button
  class="button large success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Large button
  </span>
</button>
`;

exports[`BaseButton component #size prop matches the snapshot when the size is "medium" 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Medium button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "danger" 1`] = `
<button
  class="button medium danger buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Danger button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "info" 1`] = `
<button
  class="button medium info buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Info button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "naked" 1`] = `
<button
  class="button medium naked"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Naked button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "neutral" 1`] = `
<button
  class="button medium neutral buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Neutral button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "success" 1`] = `
<button
  class="button medium success buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Success button
  </span>
</button>
`;

exports[`BaseButton component #type prop matches the snapshoot when the type is "warning" 1`] = `
<button
  class="button medium warning buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Warning button
  </span>
</button>
`;

exports[`BaseButton component With combined props matches the snapshot 1`] = `
<button
  class="button large warning fill buttonShadow"
>
  <!---->
   
  <!---->
   
  <span
    class=""
  >
    Combined button
  </span>
</button>
`;
